### PR TITLE
Integrate Multivector at segment level

### PR DESCRIPTION
**Pull request by agourlay**
This PR integrates the `MultiVector` support at the segment level.

It was tested using a version of `filtrable_hnsw_test.rs` adapted for `Multivectors`.

Two insights:

1. the accuracy of HNSW degrades as the number of vectors per multi-vectors increases
  * most likely due to the random data
  * could be validated against real data set  
2. the scoring gets expensive as the number of vectors per multi-vectors increases

### CODE REVIEW

File: lib/segment/src/common/mod.rs [hunk 1]
```diff
@@ -185,7 +185,19 @@ fn check_vector_against_config(
             Ok(())
         }
         VectorRef::Sparse(_) => Err(OperationError::WrongSparse),
-        VectorRef::MultiDense(_) => Err(OperationError::WrongMulti),
+        VectorRef::MultiDense(vectors) => {
```
Comment (agourlay) [inline lib/segment/src/common/mod.rs:188]: added a new validation

File: lib/segment/src/vector_storage/query_scorer/mod.rs [hunk 2]
```diff
@@ -29,7 +30,8 @@ pub fn score_max_similarity<TMetric: Metric<VectorElementType>>(
         sum += multi_dense_b
             .iter()
             .map(|dense_b| TMetric::similarity(dense_a, dense_b))
-            .fold(ScoreType::NEG_INFINITY, |a, b| if a > b { a } else { b });
+            .max_by(|a, b| OrderedFloat(*a).cmp(&OrderedFloat(*b)))
```
Comment (agourlay) [inline lib/segment/src/vector_storage/query_scorer/mod.rs:33]: I replaced the `fold` by a more explicit implementation using `max_by`

Also used `OrderedFloat` for consistency with our `ScoredPoint` implementation.

File: lib/segment/src/segment_constructor/segment_constructor_base.rs [hunk 3]
```diff
@@ -123,14 +124,24 @@ fn create_segment(
...
+                    open_simple_multi_dense_vector_storage(
+                        database.clone(),
+                        &db_column_name,
+                        vector_config.size,
+                        vector_config.distance,
+                        *multi_vec_config,
+                        stopped,
+                    )?
+                } else {
+                    open_simple_dense_vector_storage(
```
Comment (agourlay) [inline lib/segment/src/segment_constructor/segment_constructor_base.rs:137]: function renamed for more clarity

File: lib/segment/tests/integration/multivector_filtrable_hnsw_test.rs [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+        QueryVariant::RecommendBestScore => {
+            random_multi_vec_reco_query(rnd, dim, num_vector_per_points)
+        }
+    }
+}
+
+/// Check all cases with single vector per multi and several vectors per multi
+#[rstest]
+#[case::nearest_eq(QueryVariant::Nearest, 1, 32, 5)]
+#[case::nearest_multi(QueryVariant::Nearest, 3, 32, 20)]
```
Comment (agourlay) [inline lib/segment/tests/integration/multivector_filtrable_hnsw_test.rs:97]: this one seems to be unstable

File: lib/segment/src/fixtures/payload_fixtures.rs [hunk 5]
```diff
@@ -121,6 +121,18 @@ pub fn random_vector<R: Rng + ?Sized>(rnd_gen: &mut R, size: usize) -> DenseVect
     (0..size).map(|_| rnd_gen.gen()).collect()
 }
 
+pub fn random_multi_vector<R: Rng + ?Sized>(
+    rnd_gen: &mut R,
+    vector_size: usize,
+    num_vector_per_points: usize,
```
Comment (IvanPleshkov) [inline lib/segment/src/fixtures/payload_fixtures.rs:127]: I propose to use `1..num_vector_per_points` range and generate random mutlivector with random amount of vectors

File: lib/segment/src/vector_storage/query_scorer/mod.rs (hunk 2 above)
Comment (IvanPleshkov) [reply lib/segment/src/vector_storage/query_scorer/mod.rs:33]: Thanks! Looks much better

File: lib/segment/src/vector_storage/query_scorer/mod.rs [hunk 6]
```diff
@@ -29,7 +30,8 @@ pub fn score_max_similarity<TMetric: Metric<VectorElementType>>(
...
+            .unwrap_or(ScoreType::NEG_INFINITY);
```
Comment (IvanPleshkov) [inline lib/segment/src/vector_storage/query_scorer/mod.rs:34]: We can just `unwrap` it and add `assert` with non-empty check. `NEG_INFINITY` was for `fold` only, not needed to return it if multivector is empty

File: lib/segment/tests/integration/multivector_filtrable_hnsw_test.rs (hunk 4 above)
Comment (IvanPleshkov) [reply lib/segment/tests/integration/multivector_filtrable_hnsw_test.rs:97]: WDYM? Is it a flaky test?

File: lib/segment/tests/integration/multivector_filtrable_hnsw_test.rs [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+#[case::discovery_multi(QueryVariant::Discovery, 3, 128, 10)]
+#[case::recommend_eq(QueryVariant::RecommendBestScore, 1, 64, 5)]
+#[case::recommend_multi(QueryVariant::RecommendBestScore, 2, 64, 10)]
+fn test_multi_filterable_hnsw(
+    #[case] query_variant: QueryVariant,
+    #[case] num_vector_per_points: usize,
+    #[case] ef: usize,
+    #[case] max_failures: usize, // out of 100
+) {
+    _test_multi_filterable_hnsw(query_variant, num_vector_per_points, ef, max_failures);
```
Comment (IvanPleshkov) [inline lib/segment/tests/integration/multivector_filtrable_hnsw_test.rs:108]: Why do you need `_test_multi_filterable_hnsw` as separate function here? I propose to inline `_test_multi_filterable_hnsw`

File: lib/segment/tests/integration/multivector_filtrable_hnsw_test.rs [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+        }
+    }
+    let expected_blocks = num_points as usize / indexing_threshold * 2;
+
+    assert!(
+        (blocks.len() as i64 - expected_blocks as i64).abs() <= 3,
+        "real number of payload blocks is too far from expected"
+    );
+
+    assert_eq!(
```
Comment (IvanPleshkov) [inline lib/segment/tests/integration/multivector_filtrable_hnsw_test.rs:227]: We can remove coverage check, I guess it was copied from another test. But you can keep it if you wish

**Review by IvanPleshkov — APPROVED** [review]
Perfect! Just leaved some nitpicks

File: lib/segment/src/fixtures/payload_fixtures.rs (hunk 5 above)
Comment (agourlay) [reply lib/segment/src/fixtures/payload_fixtures.rs:127]: There is an invariant that all vectors within a point have the same dimensions.

Do you want the test using this to have points with different number of vectors maybe?

Comment (agourlay) [reply lib/segment/src/fixtures/payload_fixtures.rs:127]: Done this approach https://github.com/qdrant/qdrant/pull/3973/commits/754d10287a24a85c15d75cffb5e4244b610c078d

File: lib/segment/src/vector_storage/query_scorer/mod.rs (hunk 6 above)
Comment (agourlay) [reply lib/segment/src/vector_storage/query_scorer/mod.rs]: Done with an extra TODO to remind us to add the validation https://github.com/qdrant/qdrant/pull/3973/commits/2397628dbc159e0d7a32a821a97061affcd23930

**Pull request merged**